File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/util/TimeMeasure.java
package com.snowplowanalytics.snowplow.util;

import androidx.annotation.NonNull;

import java.io.Serializable;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

/**
 * It represents time durations and provides utility methods to convert across time units.
 */
public class TimeMeasure implements Serializable {

    /** Time duration at the selected TimeUnit. */
    public final long value;
    /** Measure unit of the time duration. */
    public final TimeUnit unit;

    /**
     * Create an object that represent a time duration at a specific time unit.
     * @param value Time duration at the selected TimeUnit.
     * @param unit Measure unit of the time duration.
     */
    public TimeMeasure(long value, @NonNull TimeUnit unit) {
        Objects.requireNonNull(unit);
        this.value = value;
        this.unit = unit;
    }

    /**
     * Convert the time unit of current time duration.
     * @param toUnit The new measure unit.
     * @return The same time duration converted to the new measure unit.
     */
    public long convert(@NonNull TimeUnit toUnit) {
        return toUnit.convert(value, unit);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TimeMeasure that = (TimeMeasure) o;
        long ms1 = convert(TimeUnit.MILLISECONDS);
        long ms2 = that.convert(TimeUnit.MILLISECONDS);

        return ms1 == ms2;
    }

    @Override
    public int hashCode() {
        int result = (int) (value ^ (value >>> 32));
        result = 31 * result + unit.hashCode();
        return result;
    }

    @NonNull
    @Override
    public String toString() {
        return "TimeMeasure{" +
                "value=" + value +
                ", unit=" + unit +
                '}';
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/util/TimeMeasure.kt
/*
 * Copyright (c) 2015-present Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.util

import java.util.concurrent.TimeUnit

/**
 * It represents time durations and provides utility methods to convert across time units.
 * @param value Time duration at the selected TimeUnit.
 * @param unit Measure unit of the time duration.
 */
class TimeMeasure(val value: Long, val unit: TimeUnit) : java.io.Serializable {

    /**
     * Convert the time unit of current time duration.
     * @param toUnit The new measure unit.
     * @return The same time duration converted to the new measure unit.
     */
    fun convert(toUnit: TimeUnit): Long {
        return toUnit.convert(value, unit)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        
        val that = other as? TimeMeasure
        val ms1 = convert(TimeUnit.MILLISECONDS)
        val ms2 = that?.convert(TimeUnit.MILLISECONDS)
        
        return ms1 == ms2
    }

    override fun hashCode(): Int {
        var result = (value xor (value ushr 32)).toInt()
        result = 31 * result + unit.hashCode()
        return result
    }

    override fun toString(): String {
        return "TimeMeasure{" +
                "value=" + value +
                ", unit=" + unit +
                '}'
    }
}
